### PR TITLE
memory_allocator

### PR DESCRIPTION
**Pull request by kilinchange**
1. 新增 LazyAllocator 类。
2. 在 graph 中创建 LazyAllocator，替代原有内存分配方案。
具体细节见设计文档：https://gxtctab8no8.feishu.cn/docx/HZcSdP6vYoP8ooxulLEc8UtBnsb?from=from_copylink

### CODE REVIEW

File: include/core/lazy_allocator.h [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    //     size: size of memory block to be freed
+    void free(size_t addr, size_t size);
+
+    // function: perform actual memory allocation
+    // return: pointer to the head address of the allocated memory
+    void *getPtr();
+
+    // function: memory alignment, rouned up
+    // return: size of the aligned memory block
+    size_t getAlignedSize(size_t size);
```
Comment (YdrMaster) [inline include/core/lazy_allocator.h:68]: 这个似乎不需要 public

**Review by YdrMaster — CHANGES_REQUESTED** [review]

File: include/core/lazy_allocator.h [hunk 2]
```diff
@@ -0,0 +1,73 @@
+#pragma once
+#include "core/runtime.h"
+#include <cstddef>
+#include <map>
+#include <unordered_set>
+
+namespace infini {
+
+class LazyAllocator {
+  protected:
```
Comment (YdrMaster) [inline include/core/lazy_allocator.h:10]: 这个这么具体的类应该不会继承了吧，而且现在也是直接以值使用的，直接 private 就行了

**Review by YdrMaster — CHANGES_REQUESTED** [review]

File: src/core/lazy_allocator.cc [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        }
+        // update the free balanced tree
+        this->freeBlocks.erase(it);
+        this->used += tailAddr - retAddr;
+    } else {
+        // the allocated memory space is not sufficient for reallocation, it
+        // needs to be extended
+        retAddr = this->peak;
+        this->peak = this->peak + size;
+        this->used += this->peak - retAddr;
```
Comment (bitzyz) [inline src/core/lazy_allocator.cc:47]: this->used是不是可以直接 += size（retAddr在这里可以干掉，把retAddr的定义放到if里面）

File: src/core/lazy_allocator.cc [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            this->tailAddrToBlockSize[retAddr + blockSize] = newBlock.blockSize;
+            this->freeBlocks.insert(newBlock);
+        }
+        // update the free balanced tree
+        this->freeBlocks.erase(it);
+        this->used += tailAddr - retAddr;
+    } else {
+        // the allocated memory space is not sufficient for reallocation, it
+        // needs to be extended
+        retAddr = this->peak;
```
Comment (bitzyz) [inline src/core/lazy_allocator.cc:45]: retAddr在if之前定义过了，这里不需要重复赋值了，后面的this->used += size就可以了吧

**Review by constroy — CHANGES_REQUESTED** [review]

File: include/core/lazy_allocator.h (hunk 2 above)
Comment (kilinchange) [reply include/core/lazy_allocator.h]: fixed

File: src/core/lazy_allocator.cc (hunk 4 above)
Comment (kilinchange) [reply src/core/lazy_allocator.cc]: fixed

**Review by YdrMaster — APPROVED** [review]

File: src/core/lazy_allocator.cc [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+namespace infini {
+LazyAllocator::LazyAllocator(Runtime runtime, size_t alignment)
+    : runtime(runtime), alignment(alignment) {
+    used = 0;
+    peak = 0;
+    ptr = nullptr;
+}
+
+LazyAllocator::~LazyAllocator() { runtime->dealloc(this->ptr); }
```
[the remark this answers is not in the document]
Comment (kilinchange) [reply src/core/lazy_allocator.cc]: fixed

File: include/core/graph.h [hunk 6]
```diff
@@ -9,9 +10,11 @@ class GraphObj : public Object {
     Runtime runtime;
     TensorVec tensors;
     OpVec ops;
+    LazyAllocator allocator;
 
   public:
-    explicit GraphObj(Runtime runtime) : runtime(runtime), sorted(false){};
+    explicit GraphObj(Runtime runtime)
+        : runtime(runtime), allocator(runtime, 8), sorted(false){};
```
[the remark this answers is not in the document]
Comment (kilinchange) [reply include/core/graph.h]: fixed.
把 allocator 的 alignment 赋值放到了 allocator 的构造函数实现里。

**Review by bitzyz — APPROVED** [review]

File: src/core/lazy_allocator.cc [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+#include <utility>
+
+namespace infini {
+LazyAllocator::LazyAllocator(Runtime runtime) : runtime(runtime) {
+    used = 0;
+    peak = 0;
+    ptr = nullptr;
+    // 'alignment' defaults to sizeof(uint64_t), because it is the length of the
+    // longest data type currently supported by the DataType field of the tensor
+    alignment = sizeof(uint64_t);
```
Comment (constroy) [inline src/core/lazy_allocator.cc:11]: 这个最好定义成一个根据设备选择的常量。CUDA 上面至少需要 256 字节对齐。

Comment (kilinchange) [reply src/core/lazy_allocator.cc]: fixed

**Review by constroy — APPROVED** [review]

**Comment by kilinchange** [comment]
Start  1: test_graph
 1/60 Test  #1: test_graph .......................   Passed    0.02 sec
      Start  2: test_graph_handler
 2/60 Test  #2: test_graph_handler ...............   Passed    0.02 sec
      Start  3: test_graph_replace
 3/60 Test  #3: test_graph_replace ...............   Passed    0.02 sec
      Start  4: test_hash
 4/60 Test  #4: test_hash ........................   Passed    0.02 sec
      Start  5: test_lazy_allocator
 5/60 Test  #5: test_lazy_allocator ..............   Passed    0.02 sec
      Start  6: test_search
 6/60 Test  #6: test_search ......................   Passed    5.89 sec
      Start  7: test_tensor_save
 7/60 Test  #7: test_tensor_save .................   Passed    0.00 sec
      Start  8: test_verify
 8/60 Test  #8: test_verify ......................   Passed    0.02 sec
      Start  9: test_batch_norm
 9/60 Test  #9: test_batch_norm ..................   Passed    0.02 sec
      Start 10: test_clip
10/60 Test #10: test_clip ........................   Passed    0.02 sec
      Start 11: test_concat
11/60 Test #11: test_concat ......................   Passed    0.02 sec
      Start 12: test_conv
12/60 Test #12: test_conv ........................   Passed    0.83 sec
      Start 13: test_conv_transposed_2d
13/60 Test #13: test_conv_transposed_2d ..........   Passed    0.02 sec
      Start 14: test_element_wise
14/60 Test #14: test_element_wise ................   Passed    0.02 sec
      Start 15: test_extend
15/60 Test #15: test_extend ......................   Passed    0.02 sec
      Start 16: test_gather
16/60 Test #16: test_gather ......................   Passed    0.02 sec
      Start 17: test_matmul
17/60 Test #17: test_matmul ......................   Passed    0.02 sec
      Start 18: test_pad
18/60 Test #18: test_pad .........................   Passed    0.02 sec
      Start 19: test_pooling
19/60 Test #19: test_pooling .....................   Passed    0.02 sec
      Start 20: test_reduce_mean
20/60 Test #20: test_reduce_mean .................   Passed    0.02 sec
      Start 21: test_reshape
21/60 Test #21: test_reshape .....................   Passed    0.02 sec
      Start 22: test_resize
22/60 Test #22: test_resize ......................   Passed    0.02 sec
      Start 23: test_slice
23/60 Test #23: test_slice .......................   Passed    0.02 sec
      Start 24: test_split
24/60 Test #24: test_split .......................   Passed    0.02 sec
      Start 25: test_bang_activation_backward
25/60 Test #25: test_bang_activation_backward ....   Passed    2.17 sec
      Start 26: test_bang_bangcKernel
26/60 Test #26: test_bang_bangcKernel ............   Passed    2.14 sec
      Start 27: test_bang_bitcompute
27/60 Test #27: test_bang_bitcompute .............   Passed    2.17 sec
      Start 28: test_bang_cast
28/60 Test #28: test_bang_cast ...................   Passed    2.16 sec
      Start 29: test_bang_ceil
29/60 Test #29: test_bang_ceil ...................   Passed    2.15 sec
      Start 30: test_bang_clip
30/60 Test #30: test_bang_clip ...................   Passed    2.25 sec
      Start 31: test_bang_concat
31/60 Test #31: test_bang_concat .................   Passed    2.24 sec
      Start 32: test_bang_conv
32/60 Test #32: test_bang_conv ...................   Passed    2.28 sec
      Start 33: test_bang_det
33/60 Test #33: test_bang_det ....................   Passed    2.15 sec
      Start 34: test_bang_element_wise
34/60 Test #34: test_bang_element_wise ...........   Passed    2.14 sec
      Start 35: test_bang_erf
35/60 Test #35: test_bang_erf ....................   Passed    2.25 sec
      Start 36: test_bang_exp
36/60 Test #36: test_bang_exp ....................   Passed    2.12 sec
      Start 37: test_bang_fill
37/60 Test #37: test_bang_fill ...................   Passed    2.14 sec
      Start 38: test_bang_floor
38/60 Test #38: test_bang_floor ..................   Passed    2.14 sec
      Start 39: test_bang_hardtanh
39/60 Test #39: test_bang_hardtanh ...............   Passed    2.21 sec
      Start 40: test_bang_l2loss
40/60 Test #40: test_bang_l2loss .................   Passed    2.16 sec
      Start 41: test_bang_log
41/60 Test #41: test_bang_log ....................   Passed    2.15 sec
      Start 42: test_bang_logic
42/60 Test #42: test_bang_logic ..................   Passed    2.25 sec
      Start 43: test_bang_matmul
43/60 Test #43: test_bang_matmul .................   Passed    2.22 sec
      Start 44: test_bang_maximum
44/60 Test #44: test_bang_maximum ................   Passed    2.16 sec
      Start 45: test_bang_minimum
45/60 Test #45: test_bang_minimum ................   Passed    2.22 sec
      Start 46: test_bang_mseloss
46/60 Test #46: test_bang_mseloss ................   Passed    2.19 sec
      Start 47: test_bang_neg
47/60 Test #47: test_bang_neg ....................   Passed    2.18 sec
      Start 48: test_bang_net
48/60 Test #48: test_bang_net ....................   Passed    2.15 sec
      Start 49: test_bang_optensor
49/60 Test #49: test_bang_optensor ...............   Passed    2.16 sec
      Start 50: test_bang_pad
50/60 Test #50: test_bang_pad ....................   Passed    2.19 sec
      Start 51: test_bang_pooling
51/60 Test #51: test_bang_pooling ................   Passed    2.17 sec
      Start 52: test_bang_pow
52/60 Test #52: test_bang_pow ....................   Passed    2.18 sec
      Start 53: test_bang_prelu
53/60 Test #53: test_bang_prelu ..................   Passed    2.15 sec
      Start 54: test_bang_reciprocal
54/60 Test #54: test_bang_reciprocal .............   Passed    2.15 sec
      Start 55: test_bang_round
55/60 Test #55: test_bang_round ..................   Passed    2.19 sec
      Start 56: test_bang_split
56/60 Test #56: test_bang_split ..................   Passed    2.23 sec
      Start 57: test_bang_sqrt
57/60 Test #57: test_bang_sqrt ...................   Passed    2.28 sec
      Start 58: test_bang_transpose
58/60 Test #58: test_bang_transpose ..............   Passed    2.26 sec
      Start 59: test_bang_trigon
59/60 Test #59: test_bang_trigon .................   Passed    2.21 sec
      Start 60: test_bang_unary
60/60 Test #60: test_bang_unary ..................   Passed    2.21 sec

100% tests passed, 0 tests failed out of 60
在  mlu 上进行了测试，测试结果全部通过。

**Comment by kilinchange** [comment]
```
      Start  1: test_graph
 1/47 Test  #1: test_graph .......................   Passed    0.03 sec
      Start  2: test_graph_handler
 2/47 Test  #2: test_graph_handler ...............   Passed    0.02 sec
      Start  3: test_graph_replace
 3/47 Test  #3: test_graph_replace ...............   Passed    0.02 sec
      Start  4: test_hash
 4/47 Test  #4: test_hash ........................   Passed    0.02 sec
      Start  5: test_lazy_allocator
 5/47 Test  #5: test_lazy_allocator ..............   Passed    0.02 sec
      Start  6: test_search
 6/47 Test  #6: test_search ......................   Passed    5.52 sec
      Start  7: test_tensor_save
 7/47 Test  #7: test_tensor_save .................   Passed    0.00 sec
      Start  8: test_verify
 8/47 Test  #8: test_verify ......................   Passed    0.02 sec
      Start  9: test_batch_norm
 9/47 Test  #9: test_batch_norm ..................   Passed    0.02 sec
      Start 10: test_clip
10/47 Test #10: test_clip ........................   Passed    0.02 sec
      Start 11: test_concat
11/47 Test #11: test_concat ......................   Passed    0.02 sec
      Start 12: test_conv
12/47 Test #12: test_conv ........................   Passed    0.03 sec
      Start 13: test_conv_transposed_2d
13/47 Test #13: test_conv_transposed_2d ..........   Passed    0.02 sec
      Start 14: test_element_wise
14/47 Test #14: test_element_wise ................   Passed    0.02 sec
      Start 15: test_extend
15/47 Test #15: test_extend ......................   Passed    0.02 sec
      Start 16: test_gather
16/47 Test #16: test_gather ......................   Passed    0.02 sec
      Start 17: test_matmul
17/47 Test #17: test_matmul ......................   Passed    0.02 sec
      Start 18: test_pad
18/47 Test #18: test_pad .........................   Passed    0.02 sec
      Start 19: test_pooling
19/47 Test #19: test_pooling .....................   Passed    0.02 sec
      Start 20: test_reduce_mean
20/47 Test #20: test_reduce_mean .................   Passed    0.02 sec
      Start 21: test_reshape
21/47 Test #21: test_reshape .....................   Passed    0.02 sec
      Start 22: test_resize
22/47 Test #22: test_resize ......................   Passed    0.02 sec
      Start 23: test_slice
23/47 Test #23: test_slice .......................   Passed    0.02 sec
      Start 24: test_split
24/47 Test #24: test_split .......................   Passed    0.02 sec
      Start 25: test_cuda_G2BMM
25/47 Test #25: test_cuda_G2BMM ..................   Passed    2.46 sec
      Start 26: test_cuda_GBMM
26/47 Test #26: test_cuda_GBMM ...................   Passed    3.33 sec
      Start 27: test_cuda_batch_norm
27/47 Test #27: test_cuda_batch_norm .............   Passed    2.24 sec
      Start 28: test_cuda_clip
28/47 Test #28: test_cuda_clip ...................   Passed    2.23 sec
      Start 29: test_cuda_concat
29/47 Test #29: test_cuda_concat .................   Passed    2.23 sec
      Start 30: test_cuda_conv
30/47 Test #30: test_cuda_conv ...................   Passed    3.87 sec
      Start 31: test_cuda_conv_fp16
31/47 Test #31: test_cuda_conv_fp16 ..............   Passed    4.13 sec
      Start 32: test_cuda_conv_transposed_2d
32/47 Test #32: test_cuda_conv_transposed_2d .....   Passed    4.28 sec
      Start 33: test_cuda_element_wise
33/47 Test #33: test_cuda_element_wise ...........   Passed    2.56 sec
      Start 34: test_cuda_extend
34/47 Test #34: test_cuda_extend .................   Passed    2.51 sec
      Start 35: test_cuda_gather
35/47 Test #35: test_cuda_gather .................   Passed    2.38 sec
      Start 36: test_cuda_inception
36/47 Test #36: test_cuda_inception ..............   Passed    3.68 sec
      Start 37: test_cuda_matmul
37/47 Test #37: test_cuda_matmul .................   Passed    2.80 sec
      Start 38: test_cuda_pad
38/47 Test #38: test_cuda_pad ....................   Passed    2.45 sec
      Start 39: test_cuda_pooling
39/47 Test #39: test_cuda_pooling ................   Passed    2.37 sec
      Start 40: test_cuda_reduce_mean
40/47 Test #40: test_cuda_reduce_mean ............   Passed    2.45 sec
      Start 41: test_cuda_reshape
41/47 Test #41: test_cuda_reshape ................   Passed    2.59 sec
      Start 42: test_cuda_resize
42/47 Test #42: test_cuda_resize .................   Passed    5.99 sec
      Start 43: test_cuda_slice
43/47 Test #43: test_cuda_slice ..................   Passed    2.53 sec
      Start 44: test_cuda_softmax
44/47 Test #44: test_cuda_softmax ................   Passed    2.53 sec
      Start 45: test_cuda_split
45/47 Test #45: test_cuda_split ..................   Passed    2.57 sec
      Start 46: test_cuda_unary
46/47 Test #46: test_cuda_unary ..................   Passed    2.41 sec
      Start 47: test_perfengine
47/47 Test #47: test_perfengine ..................   Passed    3.93 sec

100% tests passed, 0 tests failed out of 47
```
在 cuda 上进行了测试，测试结果全部通过。

**Comment by kilinchange** [comment]
```
      Start  1: test_graph
 1/40 Test  #1: test_graph .......................   Passed    0.28 sec
      Start  2: test_graph_handler
 2/40 Test  #2: test_graph_handler ...............   Passed    0.24 sec
      Start  3: test_graph_replace
 3/40 Test  #3: test_graph_replace ...............   Passed    0.28 sec
      Start  4: test_hash
 4/40 Test  #4: test_hash ........................   Passed    0.28 sec
      Start  5: test_lazy_allocator
 5/40 Test  #5: test_lazy_allocator ..............   Passed    0.27 sec
      Start  6: test_search
 6/40 Test  #6: test_search ......................   Passed    6.02 sec
      Start  7: test_tensor_save
 7/40 Test  #7: test_tensor_save .................   Passed    0.26 sec
      Start  8: test_verify
 8/40 Test  #8: test_verify ......................   Passed    0.25 sec
      Start  9: test_batch_norm
 9/40 Test  #9: test_batch_norm ..................   Passed    0.24 sec
      Start 10: test_clip
10/40 Test #10: test_clip ........................   Passed    0.23 sec
      Start 11: test_concat
11/40 Test #11: test_concat ......................   Passed    0.24 sec
      Start 12: test_conv
12/40 Test #12: test_conv ........................   Passed    0.25 sec
      Start 13: test_conv_transposed_2d
13/40 Test #13: test_conv_transposed_2d ..........   Passed    0.24 sec
      Start 14: test_element_wise
14/40 Test #14: test_element_wise ................   Passed    0.26 sec
      Start 15: test_extend
15/40 Test #15: test_extend ......................   Passed    0.25 sec
      Start 16: test_gather
16/40 Test #16: test_gather ......................   Passed    0.28 sec
      Start 17: test_matmul
17/40 Test #17: test_matmul ......................   Passed    0.27 sec
      Start 18: test_pad
18/40 Test #18: test_pad .........................   Passed    0.25 sec
      Start 19: test_pooling
19/40 Test #19: test_pooling .....................   Passed    0.25 sec
      Start 20: test_reduce_mean
20/40 Test #20: test_reduce_mean .................   Passed    0.24 sec
      Start 21: test_reshape
21/40 Test #21: test_reshape .....................   Passed    0.28 sec
      Start 22: test_resize
22/40 Test #22: test_resize ......................   Passed    0.25 sec
      Start 23: test_slice
23/40 Test #23: test_slice .......................   Passed    0.24 sec
      Start 24: test_split
24/40 Test #24: test_split .......................   Passed    0.24 sec
      Start 25: test_mkl_batch_norm
25/40 Test #25: test_mkl_batch_norm ..............   Passed    0.31 sec
      Start 26: test_mkl_concat
26/40 Test #26: test_mkl_concat ..................   Passed    0.54 sec
      Start 27: test_mkl_conv
27/40 Test #27: test_mkl_conv ....................   Passed    0.42 sec
      Start 28: test_mkl_conv_transposed
28/40 Test #28: test_mkl_conv_transposed .........   Passed    0.54 sec
      Start 29: test_mkl_element_wise
29/40 Test #29: test_mkl_element_wise ............   Passed    0.80 sec
      Start 30: test_mkl_extend
30/40 Test #30: test_mkl_extend ..................   Passed    0.92 sec
      Start 31: test_mkl_gather
31/40 Test #31: test_mkl_gather ..................   Passed    1.02 sec
      Start 32: test_mkl_matmul
32/40 Test #32: test_mkl_matmul ..................   Passed    1.04 sec
      Start 33: test_mkl_pad
33/40 Test #33: test_mkl_pad .....................   Passed    0.26 sec
      Start 34: test_mkl_pooling
34/40 Test #34: test_mkl_pooling .................   Passed    0.29 sec
      Start 35: test_mkl_reduce
35/40 Test #35: test_mkl_reduce ..................   Passed    0.31 sec
      Start 36: test_mkl_reshape
36/40 Test #36: test_mkl_reshape .................   Passed    0.29 sec
      Start 37: test_mkl_resize
37/40 Test #37: test_mkl_resize ..................   Passed    0.25 sec
      Start 38: test_mkl_slice
38/40 Test #38: test_mkl_slice ...................   Passed    0.27 sec
      Start 39: test_mkl_softmax
39/40 Test #39: test_mkl_softmax .................   Passed    0.31 sec
      Start 40: test_mkl_split
40/40 Test #40: test_mkl_split ...................   Passed    0.28 sec

100% tests passed, 0 tests failed out of 40
```
在 mkl 上进行了测试，测试结果全部通过。

**Review by whjthu — APPROVED** [review]

**Pull request merged**